.github/workflows/build-docker.yml

```diff
@@ -0,0 +1,42 @@
+name: Build Docker image
+
+on:
+  workflow_dispatch:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+
+jobs:
+  build-docker:
+    runs-on: ubuntu-24.04
+    permissions:
+      packages: write
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          submodules: recursive
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Login to GHcr.io
+        uses: docker/login-action@v1
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v6
+        with:
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+          push: ${{ github.event_name != 'pull_request' }}
```

Comment:
Instead of "push on anything that isn't pull_request", would "only on push to main" be better/safer? That way, in case someone in the future adds more stuff to the `on:` section above, that won't implicitly bypass this check.

Reply:
Seems fine. But if someone edit Dockerfile, it only tested by 2 cases:
1. Merge from user's `feature` branch, merge to `main`, then merge to this upstream server.
2. Pushed to `main` branch of this repo, as discussed.

Reply:
> it only tested by 2 cases

... unless in the future someone adds more events to the `on:` block, right? So it's arguably better to restrict this action _now_ (explicitly only run it on merge to main), rather than assume nobody will ever touch `on:`.

Reply:
<img width="1107" alt="image" src="https://github.com/user-attachments/assets/d0b25897-73eb-4b10-8099-c7e77fee5907" />

Maybe the restriction already set in the repo's setting, so we don't have to deal with the PR.

Reply:
Defense in depth (enforcing this in the GitHub workflow file _as well as_ the repo) wouldn't hurt :) But I'll let @gmarull make the final decision. I don't mean to backseat-drive this PR.